src/mono/mono/component/debugger-networking.c

```diff
@@ -0,0 +1,240 @@
+	int idx = 0;
+
+	if (!h)
+		return;
+
+	if (!info->aliases)
+		info->aliases = g_strdupv (h->h_aliases);
+
+	while (h->h_addr_list [idx]) {
+		cur = g_new0 (MonoAddressEntry, 1);
+		if (prev)
+			prev->next = cur;
+		else
+			info->entries = cur;
+
+		if (flags & MONO_HINT_CANONICAL_NAME && h->h_name)
+			cur->canonical_name = g_strdup (h->h_name);
+
+		cur->family = h->h_addrtype;
+		cur->socktype = SOCK_STREAM;
+		cur->protocol = 0; /* Zero means the default stream protocol */
+		cur->address_len = h->h_length;
+		memcpy (&cur->address, h->h_addr_list [idx], h->h_length);
+
+		prev = cur;
+		++idx;
+	}
+}
+#endif /* !defined (HAVE_GETADDRINFO) && (defined (HAVE_GETHOSTBYNAME) || defined (HAVE_GETHOSTBYNAME2)) */
+
+
+int
+mono_debugger_get_address_info (const char *hostname, int port, int flags, MonoAddressInfo **result)
+{
+#if defined (HAVE_GETADDRINFO) /* modern posix networking code */
+	char service_name [16];
+	struct addrinfo hints, *res = NULL, *info;
+	MonoAddressEntry *cur = NULL, *prev = NULL;
+	MonoAddressInfo *addr_info;
+	int ret;
+
+	memset (&hints, 0, sizeof (struct addrinfo));
+	*result = NULL;
+
+	hints.ai_family = PF_UNSPEC;
+	if (flags & MONO_HINT_IPV4)
+		hints.ai_family = PF_INET;
+	else if (flags & MONO_HINT_IPV6)
+		hints.ai_family = PF_INET6;
+
+	hints.ai_socktype = SOCK_STREAM;
+
+	if (flags & MONO_HINT_CANONICAL_NAME)
+		hints.ai_flags = AI_CANONNAME;
+	if (flags & MONO_HINT_NUMERIC_HOST)
+		hints.ai_flags |= AI_NUMERICHOST;
+
+/* Some ancient libc don't define AI_ADDRCONFIG */
+#ifdef AI_ADDRCONFIG
+	if (flags & MONO_HINT_CONFIGURED_ONLY)
+		hints.ai_flags |= AI_ADDRCONFIG;
+#endif
+	sprintf (service_name, "%d", port);
+
+	MONO_ENTER_GC_SAFE;
+	ret = getaddrinfo (hostname, service_name, &hints, &info);
+	MONO_EXIT_GC_SAFE;
+
+	if (ret)
+		return 1; /* FIXME propagate the error */
+
+	res = info;
+	*result = addr_info = g_new0 (MonoAddressInfo, 1);
+
+	while (res) {
+		cur = g_new0 (MonoAddressEntry, 1);
+		cur->family = res->ai_family;
+		cur->socktype = res->ai_socktype;
+		cur->protocol = res->ai_protocol;
+		if (cur->family == PF_INET) {
+			cur->address_len = sizeof (struct in_addr);
+			cur->address.v4 = ((struct sockaddr_in*)res->ai_addr)->sin_addr;
+#ifdef HAVE_STRUCT_SOCKADDR_IN6
+		} else if (cur->family == PF_INET6) {
+			cur->address_len = sizeof (struct in6_addr);
+			cur->address.v6 = ((struct sockaddr_in6*)res->ai_addr)->sin6_addr;
+#endif
+		} else {
+			g_warning ("Cannot handle address family %d", cur->family);
+			res = res->ai_next;
+			g_free (cur);
+			continue;
+		}
+
+		if (res->ai_canonname)
+			cur->canonical_name = g_strdup (res->ai_canonname);
+
+		if (prev)
+			prev->next = cur;
+		else
+			addr_info->entries = cur;
+
+		prev = cur;
+		res = res->ai_next;
+	}
+
+	freeaddrinfo (info);
+	return 0;
+
+#elif defined (HAVE_GETHOSTBYNAME) || defined (HAVE_GETHOSTBYNAME2) /* fallback networking code that relies on old BSD apis or whatever else is available */
```

Comment:
I wonder whether there still is any platform we care about that doesn't have `getaddrinfo`?
We could drop this code otherwise.